rotkehlchen/exchanges/woo.py

```diff
@@ -0,0 +1,431 @@
+        return True, ''
+
+    def _api_query(
+            self,
+            endpoint: str,
+            method: Literal['GET', 'POST'] = 'GET',
+            options: Optional[dict[str, Any]] = None,
+    ) -> Response:
+        """Request a  Woo API endpoint (from `endpoint`)."""
+        call_options = options if options else {}
+        request_url = f'{self.base_uri}/{endpoint}'
+        timestamp = str(ts_now_in_ms())
+        parameters = urllib.parse.urlencode(call_options)
+        normalized_content = f'{timestamp}{method}/{endpoint}{parameters}' if endpoint.startswith('v3') else f'{parameters}|{timestamp}'  # noqa: E501
+        signature = hmac.new(
+            self.secret,
+            msg=normalized_content.encode('utf-8'),
+            digestmod=hashlib.sha256,
+        ).hexdigest()
+        self.session.headers.update({
+            'x-api-signature': signature,
+            'x-api-timestamp': timestamp,
+        })
+
+        log.debug('Woo API request', request_url=request_url, options=options)
+        try:
+            response = self.session.request(
+                method=method,
+                url=request_url,
+                data=call_options if method == 'POST' else {},
+                params=call_options if method == 'GET' else {},
+                headers=self.session.headers,
+            )
+        except requests.exceptions.RequestException as e:
+            raise RemoteError(
+                f'Woo {method} request at {request_url} connection error: {e!s}.',
+            ) from e
+
+        return response
+
+    @overload
+    def _api_query_paginated(
+            self,
+            start_ts: Timestamp,
+            end_ts: Timestamp,
+            case: Literal['trades'],
+            options: dict[str, Any],
+    ) -> list[Trade]:
+        ...
+
+    @overload
+    def _api_query_paginated(
+            self,
+            start_ts: Timestamp,
+            end_ts: Timestamp,
+            case: Literal['asset_movements'],
+            options: dict[str, Any],
+    ) -> list[AssetMovement]:
+        ...
+
+    def _api_query_paginated(
+            self,
+            start_ts: Timestamp,
+            end_ts: Timestamp,
+            case: Literal['trades', 'asset_movements'],
+            options: dict[str, Any],
+    ) -> Union[list[Trade], list[AssetMovement], list]:
+        """Request a Woo API endpoint paginating via an options attribute."""
+        deserialization_method: Callable[[dict[str, Any]], Any]
+        results = []
+        endpoint = 'v1/client/hist_trades' if case == 'trades' else 'v1/asset/history'
+        call_options = {
+            'end_t': ts_sec_to_ms(end_ts),
+            'fromId': 1,
+            'limit': options.get('limit', API_MAX_LIMIT),
+            'start_t': ts_sec_to_ms(start_ts),
+        }
+        while True:
+            response = self._api_query(
+                endpoint=endpoint,
+                options=call_options,
+            )
```

Comment:
take care because this can raise remoterror. If handled higher in the call trace add it to the docs